### ISSUE TITLE
Tpetra:  `Tpetra::Details::LocalMap` cannot be default constructed on device in `ROCm`.

### ISSUE THREAD

**Issue by romintomasetti** [trilinos/Trilinos#11186]
## Bug Report
@trilinos/\<Tpetra\>

### Description

If you initialize an object on the device when compiling with `ROCm`, that in turn default-initializes a `Tpetra::Details::LocalMap`, you get compilation errors.

In our case, we build a `Kokkos::View` of such local maps on the device, and we get:
```bash
/opt/Trilinos/GNU-ROCm/include/impl/Kokkos_ViewMapping.hpp:2912:21: error: reference to __host__ function 'LocalMap' in __host__ __device__ function
      new (ptr + i) ValueType();
                    ^
/opt/Trilinos/GNU-ROCm/include/HIP/Kokkos_HIP_Parallel_Range.hpp:81:5: note: called by 'exec_range<void>'
    m_functor(i);
    ^
/opt/Trilinos/GNU-ROCm/include/HIP/Kokkos_HIP_Parallel_Range.hpp:102:22: note: called by 'operator()'
      this->template exec_range<WorkTag>(iwork);
                     ^
/opt/Trilinos/GNU-ROCm/include/HIP/Kokkos_HIP_KernelLaunch.hpp:114:11: note: called by 'hip_parallel_launch_local_memory<Kokkos::Impl::ParallelFor<Kokkos::Impl::ViewValueFunctor<Kokkos::Device<Kokkos::Experimental::HIP, Kokkos::Experimental::HIPSpace>, Tpetra::Details::LocalMap<>, false>, Kokkos::RangePolicy<Kokkos::Experimental::HIP, Kokkos::IndexType<long>>, Kokkos::Experimental::HIP>, 1024U, 1U>'
  driver->operator()();
          ^
/opt/Trilinos/GNU-ROCm/include/Tpetra_Details_LocalMap.hpp:108:3: note: 'LocalMap' declared here
  LocalMap();
  ^
1 error generated when compiling for gfx906.
```

This happens because the view's constructor tries to default-initialize the local map on device.

### Fix

The fix could be the following:
1. Change the default constructor of `Tpetra::Details::LocalMap` and how the private members are defined in `Tpetra_Details_LocalMap.hpp`:
    ```c++
   //! Default constructor.
   LocalMap() = default;
   <...>
   GlobalOrdinal indexBase_          = 0;
   GlobalOrdinal myMinGid_           = Tpetra::Details::OrdinalTraits<GlobalOrdinal>::invalid();
   GlobalOrdinal myMaxGid_           = Tpetra::Details::OrdinalTraits<GlobalOrdinal>::invalid();
   GlobalOrdinal firstContiguousGid_ = Tpetra::Details::OrdinalTraits<GlobalOrdinal>::invalid();
   GlobalOrdinal lastContiguousGid_  = Tpetra::Details::OrdinalTraits<GlobalOrdinal>::invalid();
   LocalOrdinal numLocalElements_    = 0;
   bool contiguous_                  = false;
   ```
3. Add `KOKKOS_FUNCTION` to the default constructor of `Tpetra::Details::FixedHashTable` in both the `Tpetra_Details_FixedHashTable_def.hpp` and `Tpetra_Details_FixedHashTable_decl.hpp` files.

**Comment by cgcgcg** [comment]
@trilinos/tpetra

**Comment by brian-kelley** [comment]
@romintomasetti We don't support the construction of LocalMap in device code, generally (or having Kokkos::Views of them). They are only intended to be constructed on host, and then used on host or device.

**Comment by brian-kelley** [comment]
On the other hand, I don't see any issues with the changes you proposed since the default constructor you propose is fine for device code. If you open a pull request with those changes, add me as a reviewer and I'll approve it.

**Comment by maartenarnst** [comment]
Thanks!  The fix was merged in #11218.  Feel free to close.
